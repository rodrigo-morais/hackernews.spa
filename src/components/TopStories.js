import React from 'react'
import PropTypes from 'prop-types'

const TopStories = ({ topStories: { data } }) => (
	<ul>
		{ !!data && data.map((story) => <li key={story.id}>{story.title}</li>) }
	</ul>
)

TopStories.propTypes = {
	topStories: PropTypes.object.isRequired,
}

export default TopStories
