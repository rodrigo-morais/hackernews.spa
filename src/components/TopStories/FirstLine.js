import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { HN_URL } from '../../constants'

import Tr from '../shared/Tr'
import A from '../shared/A'

const Td = styled.td`
	font-family: Verdana, Geneva, sans-serif;
	font-size: 10pt;
	color: #828282;
	vertical-align: top
`

const TdRight = styled(Td)`
	text-align: right;
`

const Vote = styled.div`
	background-size: 10px;
	background-image: url(../../../assets/images/grayarrow2x.gif);
	width: 10px;
	height: 10px;
`

const Span = styled.span`
	font-size: 10px;
	margin-left: 5px;
`

const AHost = styled(A)`
	color: #828282;
`

const FirstLine = ({ story, page, index }) => {
	const url = story.url ? new URL(story.url) : null

	return (
		<Tr>
			<TdRight>
				<span>{`${((page - 1) * 30) + (index + 1)}.`}</span>
			</TdRight>
			<Td>
				<center>
					<A href={`${HN_URL}vote?id=18014283&how=up&goto=news`}>
						<Vote />
					</A>
				</center>
			</Td>
			<Td>
				<A href={story.url}>{story.title}</A>
				{url && (
          <Span>
            (
              <AHost href={url.origin}>
                {url.hostname}
              </AHost>
            )
          </Span>
        )}
			</Td>
		</Tr>
	)
}

FirstLine.propTypes = {
	story: PropTypes.object.isRequired,
	page: PropTypes.number.isRequired,
	index: PropTypes.number.isRequired,
}

export default FirstLine
