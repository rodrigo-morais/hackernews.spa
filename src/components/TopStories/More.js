import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Tr from '../shared/Tr'
import A from '../shared/A'

const Td = styled.td`
	font-family: Verdana, Geneva, sans-serif;
	font-size: 10pt;
`

const AColor = styled(A)`
	color: #828282;
`

const More = ({ page, getNextTopStories }) => (
	<Tr>
		<Td colSpan="2"></Td>
		<Td>
			<AColor href="#" onClick={() => getNextTopStories(page + 1)}>More</AColor>
		</Td>
	</Tr>
)

More.propTypes = {
	page: PropTypes.number.isRequired,
	getNextTopStories: PropTypes.func.isRequired,
}

export default More
