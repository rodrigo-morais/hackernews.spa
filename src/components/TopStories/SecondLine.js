import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import styled from 'styled-components'

import { HN_URL } from '../../constants'

import Tr from '../shared/Tr'
import A from '../shared/A'

const Td = styled.td`
	font-family: Verdana, Geneva, sans-serif;
	font-size: 7pt;
	color: #828282;
`

const AColor = styled(A)`
	color: #828282;
`

const StyledLink = styled(Link)`
  text-decoration: none;
	color: #828282;

  &:focus, &:hover, &:visited, &:link, &:active {
    text-decoration: none;
  }
`

const SecondLine = ({ story }) => {
	const timeDiff = (new Date() - new Date(story.time * 1000)) / 1000 / 60
	const time = timeDiff > 60 ? Math.floor(timeDiff / 60) : Math.floor(timeDiff)
	const timeMeasure = `${timeDiff > 60 ? 'hour' : 'minutes'}${time === 1 ? '' : 's'}`

	return (
		<Tr>
			<Td colSpan="2"></Td>
			<Td>
				{story.score}
        &nbsp;points by&nbsp;
        <AColor href={`${HN_URL}user?id=${story.by}`}>{story.by}</AColor>
				{` ${time} ${timeMeasure} ago `}
				|&nbsp;
        <AColor href="#">hide</AColor>
        &nbsp;|
            <StyledLink to={`/item?id=${story.id}`}>
              {story.descendants === 0
                ? ' discuss'
                : ` ${story.descendants} comments`
              }
            </StyledLink>
			</Td>
		</Tr>
	)
}

SecondLine.propTypes = {
	story: PropTypes.object.isRequired,
}

export default SecondLine
