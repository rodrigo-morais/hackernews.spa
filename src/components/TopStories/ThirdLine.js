import React from 'react'
import styled from 'styled-components'

import Tr from '../shared/Tr'

const TrHeight = styled(Tr)`
	height: 5px;
`

const ThirdLine = () => (
	<TrHeight></TrHeight>
)

export default ThirdLine
