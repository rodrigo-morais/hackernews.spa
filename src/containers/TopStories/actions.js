import {
  REQUESTED_TOP_STORIES,
  RECEIVED_TOP_STORIES,
  RECEIVED_LAST_TOP_STORIES,
  FAILED_TOP_STORIES,
} from './constants'

export const requestTopStories = () => ({
  type: REQUESTED_TOP_STORIES,
})

export const receiveTopStories = (data) => ({
  type: RECEIVED_TOP_STORIES,
  data,
})

export const receiveLastTopStories = () => ({
  type: RECEIVED_LAST_TOP_STORIES,
  data: { last: true },
})

export const failTopStories = () => ({
  type: FAILED_TOP_STORIES,
})
