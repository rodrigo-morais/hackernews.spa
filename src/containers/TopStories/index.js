import { connect } from 'react-redux'

import TopStories from '../../components/TopStories'
import { getNextTopStories } from './actionsCreator'

const mapStateToProps = ({ topStories }) => ({ topStories })

const mapDispatchToProps = {
	getNextTopStories,
}

export default connect(mapStateToProps, mapDispatchToProps)(TopStories)
