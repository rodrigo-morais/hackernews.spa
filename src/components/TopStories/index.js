import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import FirstLine from './FirstLine'
import SecondLine from './SecondLine'
import ThirdLine from './ThirdLine'
import More from './More'

import TableBody from '../shared/TableBody'
import Tr from '../shared/Tr'
import Tbody from '../shared/Tbody'


const TrMoreSpace = styled(Tr)`
	height: 10px;
`

const TopStories = ({ topStories: { data: { stories, page, last } }, getNextTopStories, getComments }) => (
	<Tr>
		<td>
			<TableBody>
				<Tbody>
					{ !!stories && stories.map((story, index) => [<FirstLine story={story} page={page} key={`First_${story.id}`} index={index} />,
							<SecondLine story={story} getComments={getComments} key={`Second_${story.id}`} />,
							<ThirdLine key={`Third_${story.id}`} />])}
					<TrMoreSpace />
          {!last && <More page={page} getNextTopStories={getNextTopStories} />}
				</Tbody>
			</TableBody>
		</td>
	</Tr>
)

TopStories.propTypes = {
	topStories: PropTypes.object.isRequired,
	getNextTopStories: PropTypes.func.isRequired,
	getComments: PropTypes.func.isRequired,
}

export default TopStories
