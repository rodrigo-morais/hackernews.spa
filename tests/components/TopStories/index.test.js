import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { StaticRouter } from 'react-router-dom'

import TopStories from '../../../src/components/TopStories'
import stories from '../../data/topStories'

const DATE = new Date(1537666329761)
const _Date = Date
global.Date = jest.fn(() => DATE)
global.Date.now = _Date.now

const getNextTopStories = jest.fn()
const getComments = jest.fn()

const topStories = { data: { stories, page: 1 } }
const context = {}

describe('TopStories', () => {
  let wraper

  beforeEach(() => {
    wraper = (<StaticRouter location="/comments" context={context}>
      <TopStories topStories={topStories} getNextTopStories={getNextTopStories} getComments={getComments} />
    </StaticRouter>)
  })

  describe('Rendering', () => {
    test('renders according to design', () => {
      const component = renderer.create(wraper)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('onClick callback', () => {
    beforeEach(() => {
      getNextTopStories.mockClear();
      getComments.mockClear();
    });

    it('execute getNextTopStories callback when unlocked', () => {
      const component = mount(wraper)
      component.find('a').last().simulate('click');
      expect(getNextTopStories).toBeCalled();
    });

    it('execute getNextTopStories callback when unlocked', () => {
      const wrapper = mount(
        <StaticRouter location="/comments" context={context}>
          <TopStories topStories={topStories} getNextTopStories={getNextTopStories} getComments={getComments} />
        </StaticRouter>
      );
      wrapper.find('Link').first().simulate('click');
      expect(getComments).toBeCalled();
    });
  });
})
