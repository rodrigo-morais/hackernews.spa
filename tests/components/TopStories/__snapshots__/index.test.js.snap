// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopStories Rendering renders according to design 1`] = `
<tr
  className="sc-bdVaJa fePDys"
>
  <td>
    <table
      className="sc-VigVT ekbYvL"
    >
      <tbody
        className="sc-jTzLTM hkFqJX"
      >
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-bxivhb RPNdr sc-htpNat bBeirU"
          >
            <span>
              1.
            </span>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <center>
              <a
                className="sc-bwzfXH cQPqep"
                href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
              >
                <div
                  className="sc-ifAKCX ixPFSp"
                />
              </a>
            </center>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <a
              className="sc-bwzfXH cQPqep"
              href="https://github.com/slothking-online/graphql-editor"
            >
              GraphQL editor – Visual node editor for GraphQL
            </a>
            <span
              className="sc-EHOje fBEdJW"
            >
              (
              <a
                className="sc-bZQynM goICVT sc-bwzfXH cQPqep"
                href="https://github.com"
              >
                github.com
              </a>
              )
            </span>
          </td>
        </tr>
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-gzVnrw bmUpor"
            colSpan="2"
          />
          <td
            className="sc-gzVnrw bmUpor"
          >
            79
             points by 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="https://news.ycombinator.com/user?id=aexol"
            >
              aexol
            </a>
             0 minutess ago 
            | 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="#"
            >
              hide
            </a>
             |
            <a
              className="sc-dnqmqq eMkAIw"
              href="/item?id=18045532"
              onClick={[Function]}
            >
               3 comments
            </a>
          </td>
        </tr>
        <tr
          className="sc-iwsKbI fMAcUh sc-bdVaJa fePDys"
        />
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-bxivhb RPNdr sc-htpNat bBeirU"
          >
            <span>
              2.
            </span>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <center>
              <a
                className="sc-bwzfXH cQPqep"
                href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
              >
                <div
                  className="sc-ifAKCX ixPFSp"
                />
              </a>
            </center>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <a
              className="sc-bwzfXH cQPqep"
              href="https://github.com/manatlan/vbuild"
            >
              Vbuild: “Compile” Vue components to standalone HTML/JS/CSS
            </a>
            <span
              className="sc-EHOje fBEdJW"
            >
              (
              <a
                className="sc-bZQynM goICVT sc-bwzfXH cQPqep"
                href="https://github.com"
              >
                github.com
              </a>
              )
            </span>
          </td>
        </tr>
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-gzVnrw bmUpor"
            colSpan="2"
          />
          <td
            className="sc-gzVnrw bmUpor"
          >
            31
             points by 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="https://news.ycombinator.com/user?id=manatlan"
            >
              manatlan
            </a>
             0 minutess ago 
            | 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="#"
            >
              hide
            </a>
             |
            <a
              className="sc-dnqmqq eMkAIw"
              href="/item?id=18045813"
              onClick={[Function]}
            >
               7 comments
            </a>
          </td>
        </tr>
        <tr
          className="sc-iwsKbI fMAcUh sc-bdVaJa fePDys"
        />
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-bxivhb RPNdr sc-htpNat bBeirU"
          >
            <span>
              3.
            </span>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <center>
              <a
                className="sc-bwzfXH cQPqep"
                href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
              >
                <div
                  className="sc-ifAKCX ixPFSp"
                />
              </a>
            </center>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <a
              className="sc-bwzfXH cQPqep"
              href="https://www.polygon.com/features/2016/1/14/10757460/the-game-industry-of-iran"
            >
              The game industry of Iran (2016)
            </a>
            <span
              className="sc-EHOje fBEdJW"
            >
              (
              <a
                className="sc-bZQynM goICVT sc-bwzfXH cQPqep"
                href="https://www.polygon.com"
              >
                www.polygon.com
              </a>
              )
            </span>
          </td>
        </tr>
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-gzVnrw bmUpor"
            colSpan="2"
          />
          <td
            className="sc-gzVnrw bmUpor"
          >
            29
             points by 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="https://news.ycombinator.com/user?id=1_player"
            >
              1_player
            </a>
             0 minutess ago 
            | 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="#"
            >
              hide
            </a>
             |
            <a
              className="sc-dnqmqq eMkAIw"
              href="/item?id=18039056"
              onClick={[Function]}
            >
               discuss
            </a>
          </td>
        </tr>
        <tr
          className="sc-iwsKbI fMAcUh sc-bdVaJa fePDys"
        />
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-bxivhb RPNdr sc-htpNat bBeirU"
          >
            <span>
              4.
            </span>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <center>
              <a
                className="sc-bwzfXH cQPqep"
                href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
              >
                <div
                  className="sc-ifAKCX ixPFSp"
                />
              </a>
            </center>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <a
              className="sc-bwzfXH cQPqep"
              href="https://github.com/NVIDIA/vid2vid"
            >
              Nvidia Vid2vid: High-resolution photorealistic video-to-video translation
            </a>
            <span
              className="sc-EHOje fBEdJW"
            >
              (
              <a
                className="sc-bZQynM goICVT sc-bwzfXH cQPqep"
                href="https://github.com"
              >
                github.com
              </a>
              )
            </span>
          </td>
        </tr>
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-gzVnrw bmUpor"
            colSpan="2"
          />
          <td
            className="sc-gzVnrw bmUpor"
          >
            204
             points by 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="https://news.ycombinator.com/user?id=davedx"
            >
              davedx
            </a>
             0 minutess ago 
            | 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="#"
            >
              hide
            </a>
             |
            <a
              className="sc-dnqmqq eMkAIw"
              href="/item?id=18038084"
              onClick={[Function]}
            >
               98 comments
            </a>
          </td>
        </tr>
        <tr
          className="sc-iwsKbI fMAcUh sc-bdVaJa fePDys"
        />
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-bxivhb RPNdr sc-htpNat bBeirU"
          >
            <span>
              5.
            </span>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <center>
              <a
                className="sc-bwzfXH cQPqep"
                href="https://news.ycombinator.com/vote?id=18014283&how=up&goto=news"
              >
                <div
                  className="sc-ifAKCX ixPFSp"
                />
              </a>
            </center>
          </td>
          <td
            className="sc-htpNat bBeirU"
          >
            <a
              className="sc-bwzfXH cQPqep"
              href="https://www.nytimes.com/2018/09/22/technology/defcon-hackers-privacy-anonymity.html"
            >
              For Hackers, Anonymity Was Once Critical. That’s Changing
            </a>
            <span
              className="sc-EHOje fBEdJW"
            >
              (
              <a
                className="sc-bZQynM goICVT sc-bwzfXH cQPqep"
                href="https://www.nytimes.com"
              >
                www.nytimes.com
              </a>
              )
            </span>
          </td>
        </tr>
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-gzVnrw bmUpor"
            colSpan="2"
          />
          <td
            className="sc-gzVnrw bmUpor"
          >
            55
             points by 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="https://news.ycombinator.com/user?id=extarial"
            >
              extarial
            </a>
             0 minutess ago 
            | 
            <a
              className="sc-htoDjs flamxC sc-bwzfXH cQPqep"
              href="#"
            >
              hide
            </a>
             |
            <a
              className="sc-dnqmqq eMkAIw"
              href="/item?id=18045339"
              onClick={[Function]}
            >
               12 comments
            </a>
          </td>
        </tr>
        <tr
          className="sc-iwsKbI fMAcUh sc-bdVaJa fePDys"
        />
        <tr
          className="sc-fjdhpX XPQYG sc-bdVaJa fePDys"
        />
        <tr
          className="sc-bdVaJa fePDys"
        >
          <td
            className="sc-gZMcBi iDzNSP"
            colSpan="2"
          />
          <td
            className="sc-gZMcBi iDzNSP"
          >
            <a
              className="sc-gqjmRU gKsMen sc-bwzfXH cQPqep"
              href="#"
              onClick={[Function]}
            >
              More
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </td>
</tr>
`;
